File: src/main/java/me/nickp0is0n/easylocalize/utils/LocalizeParser.java
package me.nickp0is0n.easylocalize.utils;

import me.nickp0is0n.easylocalize.models.LocalizedString;
import me.nickp0is0n.easylocalize.models.ParserSettings;

import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class LocalizeParser {

    private String currentComment = "";
    private String currentId = null;
    private String currentString = null;
    private boolean multilineCommentMode = false;
    private String currentMark = null;
    private String header = null;
    private final ParserSettings settings;

    public LocalizeParser (ParserSettings settings) {
        this.settings = settings;
    }

    private final ArrayList<LocalizedString> strings = new ArrayList<>();

    public List<LocalizedString> fromFile(File file) throws IOException {
        BufferedReader reader = new BufferedReader(new FileReader(file));
        String currentLine;
        boolean isCommentMultilined = false;
        boolean isHeaderAlreadyExist = false;

        while ((currentLine=reader.readLine())!=null) {
            if (isAComment(currentLine)) {
                setCurrentComment(currentComment + parseComment(currentLine));
                if (strings.isEmpty() && !multilineCommentMode && !isCommentMultilined) {
                    isHeaderAlreadyExist = true;
                }
                if (multilineCommentMode) {
                    isCommentMultilined = true;
                }
                else if (!currentComment.isEmpty() && strings.isEmpty() && !isHeaderAlreadyExist){
                    header = currentComment;
                    setCurrentComment("");
                    isHeaderAlreadyExist = true;
                    isCommentMultilined = false;
                }
            }

            else if (isLineBelongToUnfinishedString(currentLine, currentString)) {
                if (settings.getIgnoreCopyrightHeader()) {
                    header = null;
                }
                parseEndOfMultilineString(currentLine, isCommentMultilined);
            }

            else if (!currentLine.equals("") || multilineCommentMode) {
                Pattern wrappedStringPattern = Pattern.compile("([\"])(?:(?=(\\\\?))\\2.)*?\\1");
                Matcher patternMatcher = wrappedStringPattern.matcher(currentLine);

                retrieveId(currentLine, patternMatcher);
                retrieveTextString(currentLine, patternMatcher);

                if (settings.getIgnoreCopyrightHeader()) {
                    header = null;
                }

                finalizeLocalizedString(currentLine, isCommentMultilined);
                isCommentMultilined = false;
            }
        }
        return strings;
    }

    private void finalizeLocalizedString(String currentLine, boolean isCommentMultilined) {
        if (currentLine.endsWith("\";")) {
            strings.add(new LocalizedString(currentId, currentString, currentComment, isCommentMultilined, currentMark, header));
            setCurrentComment("");
            header = null;
        }
    }

    private boolean isLineBelongToUnfinishedString(String currentLine, String currentString) {
        return !currentLine.startsWith("\"") && currentString != null;
    }

    private boolean isAComment(String currentLine) {
        return currentLine.startsWith("//") || currentLine.startsWith("/*") || multilineCommentMode;
    }

    private void retrieveId (String currentLine, Matcher patternMatcher) throws IOException {
        if (!patternMatcher.find()) {
            throw new IOException("ID is not found");
        }
        currentId = currentLine.substring(patternMatcher.start() + 1, patternMatcher.end() - 1);
    }

    private void retrieveTextString(String currentLine, Matcher patternMatcher) {
        if (!patternMatcher.find()) {
            int lastIndex = nthIndexOf(currentLine, '\"', 3);
            currentString = currentLine.substring(lastIndex + 1);
        }
        else {
            currentString = currentLine.substring(patternMatcher.start() + 1, patternMatcher.end() - 1);
        }
    }

    private void parseEndOfMultilineString(String currentLine, boolean isCommentMultilined) {
        //String currentString = "";
        if (currentLine.endsWith("\";")) {
            currentLine = currentLine.substring(0, currentLine.length() - 2);
            currentString = currentString + "\n" + currentLine;
            strings.add(new LocalizedString(currentId, currentString, currentComment, isCommentMultilined, currentMark, header));
            setCurrentComment("");
            header = null;
        }
        else {
            currentString = currentString + "\n" + currentLine;
        }
    }

    private String parseComment(String currentLine) {
        String currentComment = "";
        if (currentLine.startsWith("//")) {
            if (currentLine.contains("MARK:")) {
                currentMark = currentLine.substring(8);
                return currentComment;
            }
            currentComment = (currentComment + currentLine.substring(2) + "\n").trim();
            return currentComment;
        }
        if (currentLine.startsWith("/*")) {
            multilineCommentMode = true;
        }
        if (multilineCommentMode) {
            currentLine = currentLine.replace("/*", "");
            currentComment = currentComment + currentLine + "\n";
            if (currentLine.endsWith("*/")) {
                currentComment = currentComment.substring(0, currentComment.length() - 3);
                multilineCommentMode = false;
            }
        }
        return currentComment;
    }

    private int nthIndexOf(String text, char needle, int n)
    {
        for (int i = 0; i < text.length(); i++)
        {
            if (text.charAt(i) == needle)
            {
                n--;
                if (n == 0)
                {
                    return i;
                }
            }
        }
        return -1;
    }
    
    private void setCurrentComment(String comment) {
        if (settings.getIgnoreComments()) {
            currentComment = "";
        }
        else {
            currentComment = comment;
        }
    }
}

File: src/main/kotlin/me/nickp0is0n/easylocalize/utils/LocalizeParser.kt
package me.nickp0is0n.easylocalize.utils

import me.nickp0is0n.easylocalize.models.LocalizedString
import me.nickp0is0n.easylocalize.models.ParserModel
import me.nickp0is0n.easylocalize.models.ParserSettings
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.util.regex.Matcher
import java.util.regex.Pattern

class LocalizeParser(val settings: ParserSettings) {
    private val parserModel = ParserModel(settings = settings)
    private val parsedStrings = mutableListOf<LocalizedString>()

    fun fromFile(file: File): List<LocalizedString> {
        val reader = BufferedReader(FileReader(file))
        var currentLine: String?
        var isCommentMultilined = false
        var isHeaderAlreadyExist = false

        while (reader.readLine().also { currentLine = it } != null) {
            when {
                isAComment(currentLine!!) -> {
                    if (!isCommentMultilined && parserModel.currentComment.isNotEmpty()) {
                        setCurrentComment(parserModel.currentComment + "\n" + parseComment(currentLine!!))
                    } else {
                        setCurrentComment(parserModel.currentComment + parseComment(currentLine!!))
                    }

                    if (parsedStrings.isEmpty() && !parserModel.multilineCommentMode && !isCommentMultilined) {
                        isHeaderAlreadyExist = true
                    }
                    if (parserModel.multilineCommentMode) {
                        isCommentMultilined = true
                    } else if (parserModel.currentComment.isNotEmpty() && parsedStrings.isEmpty() && !isHeaderAlreadyExist) {
                        parserModel.header = parserModel.currentComment
                        setCurrentComment("")
                        isHeaderAlreadyExist = true
                        isCommentMultilined = false
                    }
                }

                isLineBelongToUnfinishedString(currentLine!!, parserModel.currentString) -> {
                    if (settings.ignoreCopyrightHeader) {
                        parserModel.header = null
                    }
                    parseEndOfMultilineString(currentLine!!, isCommentMultilined)
                }

                currentLine != "" || parserModel.multilineCommentMode -> {
                    val wrappedStringPattern = Pattern.compile("([\"])(?:(?=(\\\\?))\\2.)*?\\1")
                    val patternMatcher = wrappedStringPattern.matcher(currentLine!!)
                    retrieveId(currentLine!!, patternMatcher)
                    retrieveTextString(currentLine!!, patternMatcher)
                    if (settings.ignoreCopyrightHeader) {
                        parserModel.header = null
                    }
                    finalizeLocalizedString(currentLine!!, isCommentMultilined)
                    isCommentMultilined = false
                }
            }
        }
        return parsedStrings
    }

    private fun finalizeLocalizedString(currentLine: String, isCommentMultilined: Boolean) {
        if (currentLine.endsWith("\";")) {
            parsedStrings.add(
                LocalizedString(
                    parserModel.currentId ?: "null",
                    parserModel.currentString ?: "null",
                    parserModel.currentComment,
                    isCommentMultilined,
                    parserModel.currentMark,
                    parserModel.header
                )
            )
            setCurrentComment("")
            parserModel.header = null
        }
    }

    private fun isLineBelongToUnfinishedString(currentLine: String, currentString: String?): Boolean {
        return !currentLine.startsWith("\"") && currentString != null
    }

    private fun isAComment(currentLine: String): Boolean {
        return currentLine.startsWith("//") || currentLine.startsWith("/*") || parserModel.multilineCommentMode
    }

    private fun retrieveId(currentLine: String, patternMatcher: Matcher) {
        if (!patternMatcher.find()) {
            throw IOException("ID is not found")
        }
        parserModel.currentId = currentLine.substring(patternMatcher.start() + 1, patternMatcher.end() - 1)
    }

    private fun retrieveTextString(currentLine: String, patternMatcher: Matcher) {
        if (!patternMatcher.find()) {
            val lastIndex = nthIndexOf(currentLine, '\"', 3)
            parserModel.currentString = currentLine.substring(lastIndex + 1)
        } else {
            parserModel.currentString = currentLine.substring(patternMatcher.start() + 1, patternMatcher.end() - 1)
        }
    }

    private fun parseEndOfMultilineString(currentLine: String, isCommentMultilined: Boolean) {
        var localCurrentLine = currentLine
        if (localCurrentLine.endsWith("\";")) {
            localCurrentLine = localCurrentLine.substring(0, localCurrentLine.length - 2)
            parserModel.currentString = parserModel.currentString + "\n" + localCurrentLine
            parsedStrings.add(
                LocalizedString(
                    parserModel.currentId ?: "null",
                    parserModel.currentString ?: "null",
                    parserModel.currentComment,
                    isCommentMultilined,
                    parserModel.currentMark,
                    parserModel.header
                )
            )
            setCurrentComment("")
            parserModel.header = null
        } else {
            parserModel.currentString = parserModel.currentString + "\n" + localCurrentLine
        }
    }

    private fun parseComment(currentLine: String): String {
        var localCurrentLine = currentLine
        var currentComment = ""

        when {
            localCurrentLine.startsWith("//") -> {
                if (localCurrentLine.contains("MARK:")) {
                    parserModel.currentMark = localCurrentLine.substring(8)
                    return currentComment
                }
                currentComment = "$currentComment${localCurrentLine.substring(2)}\n".trim { it <= ' ' }
                return currentComment
            }
            localCurrentLine.startsWith("/*") -> {
                parserModel.multilineCommentMode = true
            }
        }
        if (parserModel.multilineCommentMode) {
            localCurrentLine = localCurrentLine.replace("/*", "")
            currentComment = "$currentComment$localCurrentLine\n"
            if (localCurrentLine.endsWith("*/")) {
                currentComment = currentComment.substring(0, currentComment.length - 3)
                parserModel.multilineCommentMode = false
            }
        }
        return currentComment
    }

    private fun nthIndexOf(text: String, needle: Char, n: Int): Int {
        var nIndex = n
        for (i in text.indices) {
            if (text[i] == needle) {
                nIndex--
                if (nIndex == 0) {
                    return i
                }
            }
        }
        return -1
    }

    private fun setCurrentComment(comment: String) {
        if (settings.ignoreComments) {
            parserModel.currentComment = ""
        } else {
            parserModel.currentComment = comment
        }
    }
}